api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java

```diff
@@ -112,8 +119,56 @@ private void handleTag(Criteria criteria, TokenAndListParam tags) {
 		}
 	}
 	
-	private void handleParentLocation(Criteria criteria, ReferenceAndListParam parent) {
-		handleLocationReference("loc", parent).ifPresent(loc -> criteria.createAlias("parentLocation", "loc").add(loc));
+	private void handleLocationReference(Criteria criteria, ReferenceAndListParam locationAndReferences) {
+		
+		if (locationAndReferences == null) {
+			return;
+		}
+		
+		List<ReferenceOrListParam> locationOrReference = locationAndReferences.getValuesAsQueryTokens();
+		
+		if (locationOrReference == null || locationOrReference.isEmpty()) {
+			return;
+		}
+		
+		if (locationOrReference.size() > 1) {
+			throw new IllegalArgumentException("Only one location reference is supported");
+		}
+		
+		List<ReferenceParam> locationReferences = locationOrReference.get(0).getValuesAsQueryTokens();
+		
+		if (locationReferences == null || locationReferences.isEmpty()) {
+			return;
+		}
+		
+		if (locationReferences.size() > 1) {
+			throw new IllegalArgumentException("Only one location reference is supported");
+		}
+		
+		ReferenceParam locationReference = locationReferences.get(0);
+		
+		// **NOTE: this is a *bug* in the current HAPI FHIR implementation, "below" should be the "queryParameterQualifier", not the resource type; likely need update this when/fix the HAPI FHIR implementation is fixed**
+		// this is to support queries of the type "Location?partof=below:uuid"
+		if ("below".equalsIgnoreCase(locationReference.getResourceType())) {
+			List<Criterion> belowReferenceCriteria = new ArrayList<>();
+			
+			// we need to add a join to the parentLocation for each level of hierarchy we want to search, and add "equals" criterion for each level
+			int depth = 1;
+			while (depth <= SUPPORTED_LOCATION_HIERARCHY_SEARCH_DEPTH) {
```

Comment:
Should this be a for loop?

Reply:
You mean aesthetically, or is there a bug here I'm missing?  :)  For some reason I tend to prefer while loops.

Reply:
Just aesthetics. It looks like it works.